### PR TITLE
Adding Java Unit Testing Strategy

### PR DESCRIPTION
**Pull request by Innos**
Closes issues:
Implement Java Unit Testing Strategy #1782 - SoftUni-Internal/suls-issues/issues/1782

Added a new method to the ZippedTestsParser in order to allow zipped projects to be uploaded as tests.

### CODE REVIEW

File: Open Judge System/OJS.Common/GlobalConstants.cs [hunk 1]
```diff
@@ -121,5 +122,8 @@ public static class GlobalConstants
 
         // Runtime constants
         public static readonly string JavaSourceFileExtension = $".{CompilerType.Java.GetFileExtension()}";
+
+        // Class Delimiter
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/GlobalConstants.cs]: The comment as the same as the var name, so you can safely remove it

File: Open Judge System/Web/OJS.Web.Common/ZippedTestManipulator/ZippedTestsParser.cs [hunk 2]
```diff
@@ -259,5 +263,79 @@ private static void ExtractIOIFiles(ZipFile zipFile, TestsParseResult result)
...
+        {
+            // Java Unit Testing test files
+            if (zipFile.Entries.Any(x => x.FileName.ToLower().Substring(x.FileName.Length - 7, 7) == ".in.zip"))
+            {
+                var tempDir = DirectoryHelpers.CreateTempDirectory();
+                var inputs = zipFile.EntriesSorted.Where(x => x.FileName.ToLower().Substring(x.FileName.Length - 7, 7) == ".in.zip").ToList();
+
+                foreach (var input in inputs)
+                {
+                    if (!zipFile.Entries.Any(x => x.FileName.ToLower().Contains(input.FileName.ToLower().Substring(0, input.FileName.Length - 6) + "out.zip")))
```
Comment (YanaSlavcheva) [inline Open Judge System/Web/OJS.Web.Common/ZippedTestManipulator/ZippedTestsParser.cs]: split row

File: Open Judge System/Web/OJS.Web.Common/ZippedTestManipulator/ZippedTestsParser.cs [hunk 3]
```diff
@@ -259,5 +263,79 @@ private static void ExtractIOIFiles(ZipFile zipFile, TestsParseResult result)
...
+                    {
+                        throw new ArgumentException("Невалидно име на входен тест");
+                    }
+
+                    var outputFiles = zipFile.Entries.Where(x => x.FileName.ToLower().Contains(input.FileName.ToLower().Substring(0, input.FileName.Length - 6) + "out.zip")).ToList();
```
Comment (YanaSlavcheva) [inline Open Judge System/Web/OJS.Web.Common/ZippedTestManipulator/ZippedTestsParser.cs]: split row

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs [hunk 4]
```diff
@@ -716,14 +716,14 @@ private void AddTestsToProblem(Problem problem, HttpPostedFileBase testArchive)
                 testArchive.InputStream.CopyTo(memory);
                 memory.Position = 0;
 
-                var parsedTests = ZippedTestsManipulator.Parse(memory);
+                var parsedTests = ZippedTestsParser.Parse(memory);
 
                 if (parsedTests.ZeroInputs.Count != parsedTests.ZeroOutputs.Count || parsedTests.Inputs.Count != parsedTests.Outputs.Count)
```
Comment (YanaSlavcheva) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs]: split row

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            var outputPath = $"{inputDirectoryName}\\_Compiled";
+            Directory.CreateDirectory(outputPath);
+            return outputPath;
+        }
+
+        public override string BuildCompilerArguments(string inputFolder, string outputDirectory, string additionalArguments)
+        {
+            var arguments = new StringBuilder();
+
+            // Output path argument
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            arguments.Append($"-d \"{outputDirectory}\" ");
+
+            // Additional compiler arguments
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            arguments.Append(additionalArguments);
+            arguments.Append(' ');
+
+            // Input files arguments
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            // Find compiler directory
+            var directoryInfo = new FileInfo(compilerPath).Directory;
+            if (directoryInfo == null)
+            {
+                return new CompileResult(false, $"Compiler directory is null. Compiler path value: {compilerPath}");
+            }
+
+            // Prepare process start information
+            var processStartInfo =
+                new ProcessStartInfo(compilerPath)
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: move on upper row

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                    RedirectStandardError = true,
+                    RedirectStandardOutput = true,
+                    UseShellExecute = false,
+                    WindowStyle = ProcessWindowStyle.Hidden,
+                    WorkingDirectory = directoryInfo.ToString(),
+                    Arguments = arguments
+                };
+            this.UpdateCompilerProcessStartInfo(processStartInfo);
+
+            // Execute compiler
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            var compilerOutput = ExecuteCompiler(processStartInfo);
+
+            outputDirectory = this.ChangeOutputFileAfterCompilation(outputDirectory);
+
+            // Check results and return CompilerResult instance
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            if (!Directory.Exists(outputDirectory) && !compilerOutput.IsSuccessful)
+            {
+                // Compiled file is missing
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+                return new CompileResult(false, $"Compiled file is missing. Compiler output: {compilerOutput.Output}");
+            }
+
+            if (!string.IsNullOrWhiteSpace(compilerOutput.Output))
+            {
+                // Compile file is ready but the compiler has something on standard error (possibly compile warnings)
+                return new CompileResult(true, compilerOutput.Output, outputDirectory);
+            }
+
+            // Compilation is ready without warnings
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/JavaInPlaceFolderCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/MsBuildLibraryCompiler.cs [hunk 13]
```diff
@@ -1,59 +1,54 @@
...
 
             return newOutputFile;
         }
 
-        public override string BuildCompilerArguments(string inputFile, string outputFile, string additionalArguments)
+        public override string BuildCompilerArguments(string inputFile, string outputFolder, string additionalArguments)
         {
             var arguments = new StringBuilder();
 
             // Input file argument
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/MsBuildLibraryCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/MsBuildLibraryCompiler.cs [hunk 14]
```diff
@@ -1,59 +1,54 @@
...
             arguments.Append($"\"{inputFile}\" ");
 
             // Output path argument
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/MsBuildLibraryCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.Compilers/MsBuildLibraryCompiler.cs [hunk 15]
```diff
@@ -94,6 +89,8 @@ public override CompileResult Compile(string compilerPath, string inputFile, str
                 inputFile = newInputFilePath;
             }
 
+            this.InputFile = inputFile;
+
             // Build compiler arguments
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.Compilers/MsBuildLibraryCompiler.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaProjectTestsExecutionStrategy.cs [hunk 16]
```diff
@@ -134,7 +107,15 @@ public override ExecutionResult Execute(ExecutionContext executionContext)
...
+            var compilerPath = this.GetCompilerPathFunc(executionContext.CompilerType);
+            var combinedArguments = executionContext.AdditionalCompilerArguments + this.ClassPath;
+
+            var compilerResult = this.Compile(
+                executionContext.CompilerType,
+                compilerPath,
+                combinedArguments,
+                submissionFilePath);
 
             // Assign compiled result info to the execution result
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaProjectTestsExecutionStrategy.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaProjectTestsExecutionStrategy.cs [hunk 17]
```diff
@@ -134,7 +107,15 @@ public override ExecutionResult Execute(ExecutionContext executionContext)
                 return result;
             }
 
-            var compilerResult = this.DoCompile(executionContext, submissionFilePath);
+            // Compiling
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaProjectTestsExecutionStrategy.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaProjectTestsExecutionStrategy.cs [hunk 18]
```diff
@@ -144,14 +125,16 @@ public override ExecutionResult Execute(ExecutionContext executionContext)
                 return result;
             }
 
-            var arguments = new List<string>();
-            arguments.Add(this.ClassPath);
-            arguments.Add(AdditionalExecutionArguments);
-
             // Create an executor and a checker
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaProjectTestsExecutionStrategy.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+    using OJS.Common.Models;
+    using OJS.Workers.Checkers;
+    using OJS.Workers.Common;
+    using OJS.Workers.Compilers;
+    using OJS.Workers.Executors;
+
+    public class JavaUnitTestsExecutionStrategy : JavaZipFileCompileExecuteAndCheckExecutionStrategy
+    {
+        protected const string IncorrectTestFormat =
+            "The problem's tests were not uploaded as an archive of zips. A Lecturer/Administrator should reupload the tests in the correct format.";
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: `A Lecturer/Administrator should reupload the tests in the correct format`. IMO `Reupload the tests in the correct format`. is enough. What if some day we have more roles?

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+        int failed = 0;
+        for (Result result : results){{
+            if(result.wasSuccessful()){{
+                successful += result.getRunCount();
+            }}
+            else{{
+                successful += result.getRunCount() - result.getFailureCount();
+                failed += result.getFailureCount();
+                System.out.println(result.getFailures().toString());
+            }}
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: insert blank line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+                    var directory = Path.GetDirectoryName(fileNames.Last());
+                    if (!Directory.Exists(directory) && !string.IsNullOrEmpty(directory))
+                    {
+                        Directory.CreateDirectory(directory);
+                    }
+
+                    File.WriteAllText(fileNames.Last(), projectClass);
+                }
+
+                // Compiling
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+                var compilerPath = this.GetCompilerPathFunc(executionContext.CompilerType);
+                var combinedArguments = executionContext.AdditionalCompilerArguments + this.ClassPath;
+
+                var compilerResult = this.Compile(
+                    executionContext.CompilerType,
+                    compilerPath,
+                    combinedArguments,
+                    this.WorkingDirectory);
+
+                // Modify the class path to include the folder with compiled files
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 23]
```diff
@@ -0,0 +1,319 @@
...
+                this.ClassPath = $@" -classpath ""{this.JavaLibsPath}*;{compilerResult.OutputFile}""";
+
+                // Assign compiled result info to the execution result
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: remove

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 24]
```diff
@@ -0,0 +1,319 @@
...
+
+                var testResult = this.ExecuteAndCheckTest(test, processExecutionResult, checker, message);
+                result.TestResults.Add(testResult);
+                count++;
+            }
+
+            return result;
+        }
+
+        protected override CompileResult Compile(CompilerType compilerType, string compilerPath, string compilerArguments, string submissionFilePath)
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: split row

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 25]
```diff
@@ -0,0 +1,319 @@
...
+            {
+                return new CompileResult(true, null) { OutputFile = submissionFilePath };
+            }
+
+            if (!File.Exists(compilerPath))
+            {
+                throw new ArgumentException($"Compiler not found in: {compilerPath}", nameof(compilerPath));
+            }
+
+            ICompiler compiler = Compiler.CreateCompiler(compilerType);
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: var

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs (hunk 20 above)
Comment (Innos) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: Fixed.

**Pull request merged**